firedrake/cofunction.py

```diff
@@ -168,7 +169,7 @@ def zero(self, subset=None):
         return self.assign(0, subset=subset)
 
     @PETSc.Log.EventDecorator()
-    @FunctionMixin._ad_not_implemented
+    @FunctionMixin._ad_annotate_assign
```

Comment:
I recall an argument was made in one of the weekly meetings about the annotation for cofunctions assignment being different from the one for functions assignment. I can't remember exactly the rationale behind that, but @dham might be able to provide more insights here.

Other than that, considering that this solution relies on the `FunctionAssignBlock`, which I am not sure is fully compatible with cofunctions, I would recommend also adding a test to check the TLM capabilities.

Reply:
I am going to resolve this conversation because this PR was separated from the `Cofunction.assign()` annotation issue.